test/sparse/sparse.jl

```diff
@@ -1843,3 +1843,39 @@ end
     B = A[5:-1:1, 5:-1:1]
     @test issymmetric(B)
 end
+
+@testset "optimizing sparse covariance" begin
+    n = 10
+    p = 5
+    srand(1)
+    x_sparse = sprand(n, p, .50)
+    x_dense = convert(Matrix{Float64}, x_sparse)
+    @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
+    @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
+    @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
+    @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
+
+    # Test with NaN
+    x_sparse[1,1] = NaN
+    x_dense[1,1]  = NaN
+    @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
+    @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
+    @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
+    @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
+
+    # Test with Inf
+    x_sparse[1,1] = Inf
+    x_dense[1,1]  = Inf
+    @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
+    @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
+    @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
+    @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
+
+    # Test with NaN and Inf
+    x_sparse[2,1] = NaN
+    x_dense[2,1]  = NaN
+    @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
+    @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
+    @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
+    @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
```

Comment:
Perhaps DRY out these tests? For example, something along the lines of either of these sketches
```julia
@testset "optimizing sparse covariance" begin
    srand(1)
    x_sparse = sprand(10, 5, 0.5)
    x_dense = Matrix(x_sparse)
    for case in ("orig", "with NaN", "with Inf", "NaN & Inf")
        case == "with NaN" && (x_sparse[1] = NaN; x_dense[1] = NaN)
        case == "with Inf" && (x_sparse[1] = Inf; x_dense[1] = Inf)
        case == "NaN & Inf" && (x_sparse[[1, 2]] = [NaN, Inf]; x_dense[[1, 2]] = [NaN, Inf])
        @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
        @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
        @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
        @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
end

@testset "optimizing sparse covariance" begin
    srand(1)
    x_sparse = sprand(10, 5, 0.5)
    x_dense = Matrix(x_sparse)
    for (firstentry, secondentry) in (
            (x_sparse[1], x_sparse[2]), # original matrices
            (NaN,         x_sparse[2]), # with first entry NaN
            (Inf,         x_sparse[2]), # with first entry Inf
            (NaN,         Inf) )        # with first entry NaN, second Inf
        x_sparse[[1, 2]] = [firstentry, secondentry]
        x_dense[[1, 2]] = [firstentry, secondentry]
        @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
        @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
        @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
        @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
end
```
might do the trick? :)